core/kotlinx-coroutines-core/src/internal/CustomServiceLoader.kt

```diff
@@ -0,0 +1,116 @@
+package kotlinx.coroutines.internal
+
+import java.io.*
+import java.lang.ClassCastException
+import java.net.URL
+import java.util.*
+import java.util.jar.JarFile
+import java.util.zip.ZipEntry
+import kotlin.NoSuchElementException
+import java.io.IOException
+import java.io.BufferedReader
+
+private const val PREFIX: String = "META-INF/services/"
+
+public fun <S> loadProviders(service: Class<S>, loader: ClassLoader) = CustomServiceLoader(service, loader)
+
+public class CustomServiceLoader<S> internal constructor (private val service: Class<S>, private val loader: ClassLoader?) : Iterable<S> {
```

Comment:
It shouldn't be iterable or stateful, the only API we need is `CustomServiceLoader.load(service, loader): List<..>`..
Something like 

```
object CustomServiceLoader {
   fun load(...)
}

```
is enough